qiskit_nature/second_q/problems/electronic_structure_problem.py

```diff
@@ -77,54 +76,47 @@ def filter_criterion_custom(eigenstate, eigenvalue, aux_values):
 
     """
 
-    def __init__(
-        self,
-        driver: ElectronicStructureDriver,
-        transformers: Optional[List[BaseTransformer]] = None,
-    ):
+    def __init__(self, hamiltonian: Hamiltonian) -> None:
         """
-
         Args:
-            driver: A fermionic driver encoding the molecule information.
-            transformers: A list of transformations to be applied to the driver result.
+            hamiltonian: the Hamiltonian of this problem. This needs to be an
+                :class:`.ElectronicEnergy`.
```

Comment:
this needs to be of type ElectronicEnergy? The error msg is more general right, and the instance check, in allowing sub-classes of this so it does not have to be an instance of exact class.

Does having the type ElectronicEnergy in the constructor cause an issue? That is a Hamiltonian as per the base class though it is narrowing the type. Same would apply to the other problems here. It would be helpful if it was the narrowed type since hopefully things can be detected earlier rather than some type check thrown at run time

Reply:
I think having `ElectronicEnergy` in the type hint will error in `mypy` because it will break the Liskov substitution principle (I need to double check though).

Subclasses of `ElectronicEnergy` are fine, too. Is my docstring too stringent the way its written?

Reply:
It does indeed work as expected. I updated things accordingly :+1: